main.go

```diff
@@ -163,29 +156,63 @@ func main() {
 	}
 
 	// Set a field selector so that a watch on CRDs will be limited to just the configuration policy CRD.
-	newCacheFunc := cache.BuilderWithOptions(
-		cache.Options{
-			SelectorsByObject: cache.SelectorsByObject{
-				&extensionsv1.CustomResourceDefinition{}: {
-					Field: fields.SelectorFromSet(fields.Set{"metadata.name": controllers.CRDName}),
-				},
-				&extensionsv1beta1.CustomResourceDefinition{}: {
-					Field: fields.SelectorFromSet(fields.Set{"metadata.name": controllers.CRDName}),
-				},
-			},
+	cacheSelectors := cache.SelectorsByObject{
+		&extensionsv1.CustomResourceDefinition{}: {
+			Field: fields.SelectorFromSet(fields.Set{"metadata.name": controllers.CRDName}),
 		},
-	)
+		&extensionsv1beta1.CustomResourceDefinition{}: {
+			Field: fields.SelectorFromSet(fields.Set{"metadata.name": controllers.CRDName}),
+		},
+	}
+
+	ctrlKey, err := common.GetOperatorNamespacedName()
+	if err != nil {
+		if errors.Is(err, common.ErrNoNamespace) || errors.Is(err, common.ErrRunLocal) {
+			log.Info("Running locally, skipping restrictions on the Deployment cache")
+		} else {
+			log.Error(err, "Failed to identify the controller's deployment")
+			os.Exit(1)
+		}
+	} else {
+		cacheSelectors[&appsv1.Deployment{}] = cache.ObjectSelector{
+			Field: fields.SelectorFromSet(fields.Set{
+				"metadata.namespace": ctrlKey.Namespace,
+				"metadata.name":      ctrlKey.Name,
+			}),
+		}
+	}
+
+	watchNamespace, err := common.GetWatchNamespace()
+	if strings.Contains(watchNamespace, ",") {
+		err = fmt.Errorf("multiple watched namespaces are not allowed for this controller")
+	}
+
+	if err != nil {
```

Comment:
Optional: I prefer error handling upfront since it's a more common pattern.

Reply:
Let me know what you think of the update. This is kind of a weird one, basically I wanted to add an error case, but I didn't want to change the behavior of `GetWatchNamespace` to return the error, because it was copied from somewhere else.